### PR TITLE
feat(GSDT 323): implement full onboarding & state-drivenn auth flow with NGRx

### PR DESCRIPTION
**Pull request by victorbruce**
This PR introduces the complete end-to-end authentication and user onboarding architecture, built on NgRx for state management.

It implements the full user lifecycle: **Registration** -> **Email Verification** -> **Login** -> **Onboarding Wizard (Interests/Skills)** -> **Dashboard.**

The flow is secured at every step by state-driven routing guards that check the user's UserState (`REGISTERED`, `VERIFIED`, `ACTIVE`) from the NgRx store.

1. **Core NgRx auth Store**

- Actions: Defines the full lifecycle (start, success, failure) for Login, Logout, Register, Verify Email, and Complete Onboarding.

- Reducer: Handles all state transitions, including updating the user object and isAuthenticated flag.

- Effects: Manages all API side-effects, including the simplified one-step Login (which now returns the user object) and all navigation logic (e.G., loginSuccess$ redirects to the correct route based on user.state).

- Selectors: Provides memoized selectors for all auth and onboarding states (e.g., selectIsAuthenticated, selectUser, selectIsRegistering)

2. **State-Driven Routing Guards**
Three new guards protect all application routes:

- `guest.guard.ts`: Protects `/login` & `/register`, redirecting logged-in users.

- `auth.guard.ts:` Protects the main app (e.g., `dashboard`). It now correctly checks `user.state` and redirects `VERIFIED` users to onboarding and `REGISTERED` users to verification.

- `onboarding.guard.ts`: Protects the `/onboarding/...` wizard, only allowing `VERIFIED` users to access it.

### CODE REVIEW

File: src/app/app.config.ts [hunk 1]
```diff
@@ -17,13 +17,16 @@ import { UIEffects } from './store/ui/ui.effects';
...
+import { AuthEffects } from './store/auth/auth.effects';
+
 export const appConfig: ApplicationConfig = {
   providers: [
     provideBrowserGlobalErrorListeners(),
     provideRouter(routes),
     provideHttpClient(withInterceptors([globalHttpErrorInterceptor])),
-    provideStore({ ui: uiReducer }),
-    provideEffects([UIEffects]),
+    provideStore({ ui: uiReducer, auth: authReducer }),
```
Comment (Kwabena-Owusu-Achiaw) [inline src/app/app.config.ts]: create a separate file and place all your app state in it

File: src/app/core/guards/auth-guard.ts [hunk 2]
```diff
@@ -1,17 +1,37 @@
...
+  const { APP_ROUTES } = APP_CONSTANTS;
 
-  const isAuthenticated = authService.isAuthenticated();
+  return store.select(AuthSelectors.selectCurrentUser).pipe(
+    take(1),
+    map((user) => {
+      if (!user) {
+        router.navigateByUrl(APP_ROUTES.LOGIN);
+        return false;
+      }
```
Comment (Kwabena-Owusu-Achiaw) [inline src/app/core/guards/auth-guard.ts]: lets use a selectSignal for this to avoid subscription

File: src/app/core/guards/auth-guard.ts [hunk 3]
```diff
@@ -1,17 +1,37 @@
...
-    return true;
-  }
-
-  router.navigate(['/login'], { queryParams: { returnUrl: state.url } });
-  return false;
+      switch (user.state) {
+        case UserState.ACTIVE:
+          return true;
+        case UserState.VERIFIED:
+          return true;
```
Comment (Kwabena-Owusu-Achiaw) [inline src/app/core/guards/auth-guard.ts]: combine  Userstate.ACTIVE case and UserState.VERIFIED since they all return the same thing. this will give it a cleaner feel.

File: src/app/core/guards/guest-guard.ts [hunk 4]
```diff
@@ -1,15 +1,25 @@
...
-    router.navigateByUrl('/dashboard');
-    return false;
-  }
+  return store.select(AuthSelectors.selectIsAuthenticated).pipe(
+    take(1),
+    map((isAuthenticated) => {
+      if (isAuthenticated) {
+        router.navigateByUrl(APP_ROUTES.DASHBOARD);
+        return false;
+      }
```
Comment (Kwabena-Owusu-Achiaw) [inline src/app/core/guards/guest-guard.ts]: Let us use a select signal here

File: src/app/core/guards/onboarding-guard.ts [hunk 5]
```diff
@@ -1,22 +1,38 @@
...
-    router.navigate(['/login'], { queryParams: { returnUrl: state.url } });
-    return false;
-  }
+  return store.select(AuthSelectors.selectCurrentUser).pipe(
+    take(1),
+    map((user) => {
+      if (!user) {
+        router.navigateByUrl(APP_ROUTES.LOGIN);
+        return false;
+      }
```
Comment (Kwabena-Owusu-Achiaw) [inline src/app/core/guards/onboarding-guard.ts]: Let us use a select signal here

File: src/app/core/models/auth.model.ts [hunk 6]
```diff
@@ -35,25 +38,43 @@ export enum UserState {
...
+}
+
+export interface CompleteOnboardingRequest {
+  skills: UserSkill[];
+}
 
-export type VerificationSuccessResponse = ApiResponse<RegistrationData>;
+export type RegistrationSuccessResponse = ApiResponse<User>;
+export type VerificationSuccessResponse = ApiResponse<User>;
+export type LoginSuccessResponse = ApiResponse<User>;
```
Comment (Kwabena-Owusu-Achiaw) [inline src/app/core/models/auth.model.ts]: refactor these types, they are all the same.

File: src/app/core/services/auth/auth-service.ts [hunk 7]
```diff
@@ -23,16 +26,24 @@ export class AuthService {
...
 
-  public login(payload: LoginRequest): Observable<AuthResponse> {
-    return this.api.post<AuthResponse>(API_ENDPOINTS.LOGIN, payload);
+  public login(payload: LoginRequest): Observable<LoginSuccessResponse> {
+    return this.api.post<LoginSuccessResponse>(API_ENDPOINTS.LOGIN, payload);
   }
 
-  public verifyOtp(code: string, email: string): Observable<VerificationSuccessResponse> {
-    const params = new HttpParams().set('code', code).set('email', email);
+  public verifyEmail(payload: VerifyEmailRequest): Observable<VerificationSuccessResponse> {
```
Comment (Kwabena-Owusu-Achiaw) [inline src/app/core/services/auth/auth-service.ts]: destructure the values in payload for a cleaner look

File: src/app/core/services/onboarding/onboarding-data-service.spec.ts [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+
+  beforeEach(() => {
+    TestBed.configureTestingModule({});
+    service = TestBed.inject(OnboardingDataService);
+  });
+
+  it('should be created', () => {
+    expect(service).toBeTruthy();
+  });
+});
```
Comment (Kwabena-Owusu-Achiaw) [inline src/app/core/services/onboarding/onboarding-data-service.spec.ts:106]: let's add more test cases here

File: src/app/store/auth/auth.effects.ts [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+      ofType(AuthActions.registerUser),
+      switchMap(
+        ({ request }) =>
+          this.authService.register(request).pipe(
+            map((response) => AuthActions.registerUserSuccess({ user: response.data })),
+            catchError((httpError: HttpErrorResponse) => {
+              const appError = this.errorHandlerService.getError(httpError);
+              return of(AuthActions.registerUserFailure({ error: appError }));
+            }),
+          ) as Observable<Action>,
```
Comment (Kwabena-Owusu-Achiaw) [inline src/app/store/auth/auth.effects.ts]: this casting can be avoided

File: src/app/store/auth/auth.effects.ts [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+import { AuthService, ErrorHandlerService, APP_CONSTANTS, UserState } from '@app/core';
+import { HttpErrorResponse } from '@angular/common/http';
+
+const { APP_ROUTES, FULL_PAGE_ROUTES } = APP_CONSTANTS;
+@Injectable()
+export class AuthEffects {
+  private actions$ = inject(Actions);
+  private authService = inject(AuthService);
+  private errorHandlerService = inject(ErrorHandlerService);
+  private router = inject(Router);
```
Comment (Kwabena-Owusu-Achiaw) [inline src/app/store/auth/auth.effects.ts]: let us use  a constructor to instantiate these properties

File: src/app/store/auth/auth.effects.ts [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+    ),
+  );
+
+  public login$ = createEffect(() =>
+    this.actions$.pipe(
+      ofType(AuthActions.login),
+      switchMap(
+        ({ request }) =>
+          this.authService.login(request).pipe(
+            map((response) => AuthActions.loginSuccess({ user: response.data })),
```
Comment (Kwabena-Owusu-Achiaw) [inline src/app/store/auth/auth.effects.ts]: destructure data from response

File: src/app/store/auth/auth.selectors.ts
Comment (Kwabena-Owusu-Achiaw) [inline src/app/store/auth/auth.selectors.ts:1]: let us remove the undefined option at  the point of usage, the selected valued will be defined

File: src/app/core/guards/onboarding-guard.ts (hunk 5 above)
Comment (victorbruce) [reply src/app/core/guards/onboarding-guard.ts]: Rectified 👍

File: src/app/core/services/onboarding/onboarding-data-service.spec.ts (hunk 8 above)
Comment (victorbruce) [reply src/app/core/services/onboarding/onboarding-data-service.spec.ts:106]: More test cases have been added to test edge cases extensively

File: src/app/core/services/api/api-service.spec.ts [hunk 12]
```diff
@@ -28,7 +28,7 @@ describe('ApiService', () => {
     it('should make a GET request with correct URL', () => {
       const testData = { id: 1, name: 'Test' };
 
-      service.get<{id: number; name: string;}>('/test').subscribe((data) => {
+      service.get<any>('/test').subscribe((data) => {
```
Comment (Kwabena-Owusu-Achiaw) [inline src/app/core/services/api/api-service.spec.ts]: Let us provide the right type here

Comment (victorbruce) [reply src/app/core/services/api/api-service.spec.ts]: This was a mistake. The issue has been resolved

**Review by Kwabena-Owusu-Achiaw — APPROVED** [review]
